# ITU DATABEES - AI & DATA CREW HOMEWORK 2
### Complete all the tasks and send it back before the deadline. Write clear codes and gain as much point as you can. If needed, please fill the "Answer" parts in questions, otherwise keep it blank. For your educational benefit, AI usage is NOT recommended, try to complete the tasks without using any AI tools. Correct and effective codes will be sent after the deadline.
### Note: Bonus questions are not compulsory, but I recommend to deal with them also.

### Good Luck.

### Ömer Türk

In [1]:
##########################
#### Name: Ömer
#### Surname: Türk
##########################

## Mission of this homework
In this notebook, we are going to work on a student performance data. Our objective is to understand the data, perform some operations and gain insights from the dataset, facilitating our job for next works.

#### Importing Libraries

In [2]:
import pandas as pd
import numpy as np

#### Load Dataset

In [3]:
df = pd.read_csv("StudentPerformanceFactors.csv")
df.head()

,Hours_Studied,Attendance,Parental_Involvement,Access_to_Resources,Extracurricular_Activities,Sleep_Hours,Previous_Scores,Motivation_Level,Internet_Access,Tutoring_Sessions,Family_Income,Teacher_Quality,School_Type,Peer_Influence,Physical_Activity,Learning_Disabilities,Parental_Education_Level,Distance_from_Home,Gender,Exam_Score
0,23,84,Low,High,No,7,73,Low,Yes,0,Low,Medium,Public,Positive,3,No,High School,Near,Male,67
1,19,64,Low,Medium,No,8,59,Low,Yes,2,Medium,Medium,Public,Negative,4,No,College,Moderate,Female,61
2,24,98,Medium,Medium,Yes,7,91,Medium,Yes,2,Medium,Medium,Public,Neutral,4,No,Postgraduate,Near,Male,74
3,29,89,Low,Medium,Yes,8,98,Medium,Yes,1,Medium,Medium,Public,Negative,4,No,High School,Moderate,Male,71
4,19,92,Medium,Medium,Yes,6,65,Medium,Yes,3,Medium,High,Public,Neutral,4,No,College,Near,Female,70


## Section 1: Introducing Data & Basic Statistics

### Task 1: How many rows and columns on the dataset?

In [4]:
df.shape

(6607, 20)

Answer: 6607 rows and 20 columns

### Task 2: Is there any empty/null value on the dataset? If so, which column(s) and how many empty value for each?

In [5]:
df.isnull().sum()

Hours_Studied                  0
Attendance                     0
Parental_Involvement           0
Access_to_Resources            0
Extracurricular_Activities     0
Sleep_Hours                    0
Previous_Scores                0
Motivation_Level               0
Internet_Access                0
Tutoring_Sessions              0
Family_Income                  0
Teacher_Quality               78
School_Type                    0
Peer_Influence                 0
Physical_Activity              0
Learning_Disabilities          0
Parental_Education_Level      90
Distance_from_Home            67
Gender                         0
Exam_Score                     0
dtype: int64

Answer:

### Task 3: What is the average, maximum and minimum values of exam scores?

In [7]:
print("Max:",df["Exam_Score"].max())
print("Min:",df["Exam_Score"].min())
print("Avg:",df["Exam_Score"].mean())

Max: 101
Min: 55
Avg: 67.23565914938702


Answer:

### Task 4: What is the standart deviation of students' study hours?

In [8]:
df['Hours_Studied'].std()

5.990594336578764

Answer:

### Task 5: Demonstrate the numerical information for numeric columns.
Hint: use describe()

In [9]:
df.describe()

,Hours_Studied,Attendance,Sleep_Hours,Previous_Scores,Tutoring_Sessions,Physical_Activity,Exam_Score
count,6607.000000,6607.000000,6607.00000,6607.000000,6607.000000,6607.000000,6607.000000
mean,19.975329,79.977448,7.02906,75.070531,1.493719,2.967610,67.235659
std,5.990594,11.547475,1.46812,14.399784,1.230570,1.031231,3.890456
min,1.000000,60.000000,4.00000,50.000000,0.000000,0.000000,55.000000
25%,16.000000,70.000000,6.00000,63.000000,1.000000,2.000000,65.000000
50%,20.000000,80.000000,7.00000,75.000000,1.000000,3.000000,67.000000
75%,24.000000,90.000000,8.00000,88.000000,2.000000,4.000000,69.000000
max,44.000000,100.000000,10.00000,100.000000,8.000000,6.000000,101.000000


## Section 2: Data Filtering and Grouping

### Task 1: Assume that students who studies over 25 hours are "hardworking", while those who studies less than 15 hours are "lazy". What is the averages of previous exam scores for these groups?

In [14]:
hardworking = df[df['Hours_Studied'] > 25]
lazy = df[df['Hours_Studied'] < 15]
hw_avg = hardworking['Previous_Scores'].mean()
lz_avg = lazy['Previous_Scores'].mean()
print("hardworking average:",hw_avg)
print("lazy average:",lz_avg)

hardworking average: 76.03324808184144
lazy average: 74.87142857142857


Answer:

### Task 2: Calculate the average exam scores of students whose Parental_Involvement level is 'High' and students who are 'Low'.

In [15]:
high_pi_avg = df[df['Parental_Involvement'] == 'High']['Exam_Score'].mean()
low_pi_avg = df[df['Parental_Involvement'] == 'Low']['Exam_Score'].mean()
print("High:",high_pi_avg)
print("Low:",low_pi_avg)

High: 68.09276729559748
Low: 66.35826477187734


Answer:

### Task 3: What is the numerical distribution of students studying in private schools and students studying in public schools? which is more numerous?

In [17]:
print(df['School_Type'].value_counts())

School_Type
Public     4598
Private    2009
Name: count, dtype: int64


Answer:

### Task 4: What is the education level of the parents of the student with the lowest grade?

In [18]:
min_score = df['Exam_Score'].min()
lowest_parents_edu = df[df['Exam_Score'] == min_score]['Parental_Education_Level'].values[0]
lowest_parents_edu

'High School'

Answer:

### Task 5: Analyze the top 11 highest-scoring students: Does their proximity to school (Distance_from_Home) show a specific trend? Do they predominantly live 'Near', 'Moderate' or 'Far'?

In [22]:
top_11 = df.nlargest(11, 'Exam_Score')
distance_trend = top_11['Distance_from_Home'].value_counts()
distance_trend

Distance_from_Home
Near        8
Moderate    2
Far         1
Name: count, dtype: int64

Answer:

### Task 6: Print a table which demonstrates the distribution of motivation level and internet access.
Hint: use pd.crosstab()

In [23]:
pd.crosstab(df['Motivation_Level'], df['Internet_Access'])

Internet_Access,No,Yes
Motivation_Level,,
High,92,1227
Low,162,1775
Medium,245,3106


## Section 3: Bonus Questions

### Task 1: Which gender is attending to classes more? Male or Female?

In [27]:
df.groupby('Gender')['Attendance'].mean()

Gender
Female    79.868958
Male      80.056896
Name: Attendance, dtype: float64

Answer:

### Task 2: Is there a relation between teacher quality and student attendance? Do students who has better teachers attend more?

In [28]:
df.groupby('Teacher_Quality')['Attendance'].mean()

Teacher_Quality
High      80.047252
Low       80.251142
Medium    79.936306
Name: Attendance, dtype: float64

Answer: